synapse/handlers/room_member.py

```diff
@@ -711,8 +744,78 @@ def _lookup_3pid(self, id_server, medium, address):
 
         except IOError as e:
             logger.warn("Error from identity server lookup: %s" % (e,))
+
+        return None
+
+    @defer.inlineCallbacks
+    def _lookup_3pid_v2(self, id_server, medium, address, hash_details):
+        """Looks up a 3pid in the passed identity server using v2 lookup.
+
+        Args:
+            id_server (str): The server name (including port, if required)
+                of the identity server to use.
+            medium (str): The type of the third party identifier (e.g. "email").
+            address (str): The third party identifier (e.g. "foo@example.com").
+            hash_details (dict[str, str]): A dictionary containing hashing information
+                provided by an identity server.
+
+        Returns:
+            str: the matrix ID of the 3pid, or None if it is not recognized.
+        """
+        # Extract information from hash_details
+        supported_lookup_algorithms = hash_details["algorithms"]
+        lookup_pepper = hash_details["lookup_pepper"]
+
+        # Check if none of the supported lookup algorithms are present
+        if not any(
+            i in supported_lookup_algorithms
+            for i in [LookupAlgorithm.SHA256, LookupAlgorithm.NONE]
+        ):
```

Comment:
its probably clearer to do:

```python
their_algorithms = set(hash_details["algorithms"])
our_algorithms = {LookupAlgorithm.SHA256, LookupAlgorithm.NONE}

if not their_algorithms & our_algorithms:
    return

...
```

Or in fact, just don't have this check and simply handle the `else` case below